### PR TITLE
Refactor type system: added pointer, i1->u1, created data_type.inc.h

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue id = #577 #395

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: taichi/transforms/type_check.cpp [hunk 1]
```diff
@@ -323,8 +327,17 @@ class TypeCheck : public IRVisitor {
     stmt->ret_type = VectorType(1, DataType::i32);
   }
 
+  void visit(GetRootStmt *stmt) {
+    stmt->ret_type = VectorType(1, DataType::gen, true);
+  }
+
+  void visit(SNodeLookupStmt *stmt) {
+    stmt->ret_type = VectorType(1, DataType::gen, true);
```
Comment (archibate) [inline taichi/transforms/type_check.cpp]: Are we begining to make var Vector(1)?

File: taichi/transforms/type_check.cpp [hunk 2]
```diff
@@ -94,6 +96,7 @@ class TypeCheck : public IRVisitor {
   }
 
   void visit(GlobalPtrStmt *stmt) {
+    stmt->ret_type.set_is_pointer(true);
```
Comment (archibate) [inline taichi/transforms/type_check.cpp]: Great! We now distinguish ptr from i32! But their size doesn't same, also note that global_ptr's addr can overlaps external_ptr's.

File: misc/format_server.py [hunk 3]
```diff
@@ -30,8 +30,10 @@ def exec(self, cmd):
     def render_index(self):
         pulls = requests.get(
             f'https://api.github.com/repos/taichi-dev/taichi/pulls?state=open').json()
-        self.writeln(f'Click to auto-format PR')
+        self.writeln(f'Click to auto-format PR. <b>[Please do not click if the PR is not owned/reviewed by you.]</b>')
```
Comment (archibate) [inline misc/format_server.py]: They can if they rude, we may want to fliter PRs for current login user, may detect cookie.

Comment (archibate) [reply misc/format_server.py]: Also an administrator should have access to all PRs, how about to intergrate this into github actions?

File: taichi/transforms/type_check.cpp (hunk 2 above)
Comment (yuanming-hu) [reply taichi/transforms/type_check.cpp]: Yeah. I'm just implementing a temporary solution before we move to a more delicate type system :-)

File: misc/format_server.py (hunk 3 above)
Comment (yuanming-hu) [reply misc/format_server.py]: Please see #621 :-)

File: taichi/transforms/type_check.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/transforms/type_check.cpp]: No need to modify anything. This is just a backward-compatible and temporary enhancement to the type system.

**Pull request merged**